model/model.go

```diff
@@ -131,6 +137,63 @@ type LocationInfo struct {
 	ResolverIP string
 }
 
+// PrivacySettings contains privacy settings for submitting measurements.
+type PrivacySettings struct {
+	// IncludeASN indicates whether to include the ASN
+	IncludeASN bool
+
+	// IncludeCountry indicates whether to include the country
+	IncludeCountry bool
+
+	// IncludeIP indicates whether to include the IP
+	IncludeIP bool
+}
+
+// Apply applies the privacy settings to the |m| measurement using the
+// information inside of |li| to perform scrubbing.
+func (ps PrivacySettings) Apply(m *Measurement, li LocationInfo) (err error) {
+	if ps.IncludeASN == false {
+		m.ProbeASN = DefaultProbeASNString
+	}
+	if ps.IncludeCountry == false {
+		m.ProbeCC = DefaultProbeCC
+	}
+	if ps.IncludeIP == false {
+		m.ProbeIP = DefaultProbeIP
+		err = ps.maybeRewriteTestKeys(m, li.ProbeIP)
+	}
+	return
+}
+
+func (ps PrivacySettings) maybeRewriteTestKeys(m *Measurement, currentIP string) error {
+	if net.ParseIP(currentIP) == nil {
+		return errors.New("Invalid probe IP string")
+	}
+	data, err := json.Marshal(m.TestKeys)
+	if err != nil {
+		return err
+	}
+	// The check using Count is to save an unnecessary copy performed by
+	// ReplaceAll when there are no matches into the body. This is what
+	// we would like the common case to be, meaning that the code has done
+	// its job correctly and has not leaked the IP.
+	bpip := []byte(currentIP)
+	if bytes.Count(data, bpip) <= 0 {
+		return nil
+	}
+	// Killing the previous TestKeys in my intention reduces the memory
+	// pressure possibly caused by this operation. We have, in fact, two
+	// copies of the `data` slice plus the test keys. TBH, I'm not sure
+	// however that losing all references to the test keys here is enough
+	// that they are immediately cleared, or whether that happens later.
+	m.TestKeys = nil
```

Comment:
As mentioned in the previous round you don't need to do this. The golang garbage collector will reap this object once it goes out of scope (i.e. when you do the assignment below).

Now that I re-read the code I actually think it's a bit dangerous to do this `nil` assignment before you do the `json.Unmarshal` because in case that call fails you have deleted the `TestKeys`.

Reply:
Fixed in 2550506